Update BuildTools

joperezr opened the pull request:
Updating BuildTools to fix issue with versioning where some packages use '.' instead of '-'

cc: @maririos @jhendrixMSFT @chcosta

stephentoub commented:
LGTM

joperezr commented:
0994c5e contains changes that will also be required for versioning to work correctly in our VSO official builds.

ellismg commented on build.sh:
```diff
@@ -112,7 +112,7 @@ prepare_native_build()
     # Generate version.c if specified, else have an empty one.
     __versionSourceFile=$__scriptpath/bin/obj/version.c
     if [ $__generateversionsource == true ]; then
-        $__scriptpath/Tools/corerun $__scriptpath/Tools/MSBuild.exe "$__scriptpath/build.proj" /t:GenerateVersionSourceFile /p:NativeVersionSourceFile=$__scriptpath/bin/obj/version.c /p:GenerateVersionSourceFile=true /v:minimal
```
My gut feel is that passing all the unprocessed arguments to this msbuild invocation is going to be problematic or confusing long term.  I would expect the extra arguments I pass to build.sh to only flow to the msbuild for build.proj. Maybe we should teach build.sh and build.cmd about a special argument which specifies we are doing an official build and then passes the argument to msbuild here as necessary?

Interested in how other folks feel about this.

joperezr replied:
I can take this approach if you want me to. Given that we are only really running one target in this msbuild invocation, I don't really think there would be any difference as there are hardly any properties that this target relies on. Again, if you feel strongly about it, I can totally move this to a param instead.

chcosta replied:
@maririos may want to chime in on this. 

I,  personally,  like the discoverability (via usage)  when arguments are directly supported in these wrapper scripts, but maintenance and consistency across scripts can be a bit of an annoyance though and problematic if left unsupervised. Ie,  are we supporting an official build switch in every script? What else are we making a part of the script vs some hidden property? 

If we can't be consistent or very intentional about what is supported in these wrapper scripts,  then I'd prefer to leave them extremely basic and provide documentation via some other method (print a link to an MD file with usage perhaps).

maririos replied:
When I'm reading the scripts having arguments helps me to understand what is happening. This arguments by themselves are not clear and if we need them all together in order to work I prefer to have them encapsulated in a property.

What we've been trying to do is that if we need to pass just a property to a script in order to activate something then there is no need to create an argument for the script, we will just have documentation about it. But if we need a set of properties/targets/etc (like this case) create an argument is preferred. Is that correct @weshaggard ?

Currently I'm working on a common setup script (cmd/sh) that will encapsulate all the properties we need across the dev workflow scripts, so if we found that a "build official" property needs to be set in other scripts, we could defined it in the common setup script.

weshaggard replied:
At a high level we want to keep most of the heavy lifting inside our build engine and less in the scripts. We should only pass the minimal amount of arguments to msbuild in order to accomplish the task at hand. 

For this instance `/t:GenerateVersionSourceFile /p:NativeVersionSourceFile=$__scriptpath/bin/obj/version.c /p:GenerateVersionSourceFile=true` there does seem to be a bit of redundancy that doesn't feel like it is needed. It would seem as though we could get away with simply passing `/t:GenerateVersionSourceFile` and the path could be defaulted in the targets file and it doesn't seem like we really need the property if we are explicitly calling this target.

As for figuring out what needs to be passed I think we can still specify that information in the script usage and link to md files for the overall documentation for the dev workflow.

joperezr replied:
That makes sense. Although to fix all that we will require an update in buildtools first. If it's ok with you Wes, I would like to get Joel unblocked and get this in, and then fix this in buildtools and update this scripts.

weshaggard replied:
Yes lets keep people unblocked but make sure we have the work tracked in an issue.

stephentoub commented:
Test Innerloop Windows_NT Debug Build and Test please

joperezr commented:
@dotnet-bot test Innerloop Windows_NT Debug Build and Test

maririos commented on src/Native/Windows/build-native.cmd:
```diff
@@ -112,7 +116,7 @@ exit /b 1
 :GenVSSolution
 :: Generate Version file
-msbuild /t:GenerateVersionHeader build.proj /p:NativeVersionHeaderFile="%__binDir%\obj\_version.h" /p:GenerateVersionHeader=true
+msbuild /t:GenerateVersionHeader build.proj /p:NativeVersionHeaderFile="%__binDir%\obj\_version.h" /p:GenerateVersionHeader=true %OfficialBuildIdArg%
```
Could we be consistent in where this properties are being passed? I don't know if it is possible but if we are passing them in build.sh we should pass them in build.cmd too for consistency

weshaggard replied:
agreed I would expect this to be needed in both places.

weshaggard commented on src/Native/Windows/build-native.cmd:
```diff
@@ -29,6 +30,9 @@ if /i [%1] == [/p:Platform]    (
     echo Error: Invalid platform args "%1 and %2"
     exit /b 1
 )
+if /i [%1] == [/p:OfficialBuildId] (
```
Why do you need to parse this in the cmd file directly as opposed to just passing it through to msbuild?

joperezr commented:
I have addressed all feedback except for the comment Wes made which requires changes in BuildTools first. If people are ok with this, I would like to push it to unblock @jhendrixMSFT

weshaggard commented on build.cmd:
```diff
@@ -8,9 +8,46 @@ setlocal
+    goto Next
+)
+
+if /I [%1] == [officialbuildid] (
```
Can we still make this `[/p:OfficalBuildId]`? I would like to make the way we pass this consistent.

joperezr replied:
Done.

weshaggard commented:
LGTM Please squash and get it merged and make sure we have tracking issued to clean this up some more.

stephentoub commented:
> Please squash

FYI, this can be done now when merging.  When you click the Merge pull request green button, you have two options, one of which is squashing instead of creating a merge commit.  Really nice new feature.

joperezr commented:
@stephentoub Thanks for the tip, I did it, it's awesome :smile:

weshaggard commented:
@stephentoub do you like that new feature? I'm curious how it merges comments of the commits. This squashed commit didn't have any good comments (@joperezr something you should work on for the future :)) but for commits that have comments I'm not sure how the get merged together.

stephentoub commented:
> do you like that new feature?

I do, very much.

> I'm curious how it merges comments of the commits.

It gives you an edit box that initially contains all of the commits' text appended, but you can edit it however you like before hitting the merge button.  The workflow is simply: click the big green merge button, edit the text however you like, merge.

weshaggard commented:
@stephentoub nice I will have to try it out.
